rclpy/rclpy/node.py

```diff
@@ -233,41 +316,163 @@ def get_parameter(self, name: str) -> Parameter:
+        :return: Requested parameter, or alternative value if it hadn't been declared before.
+        """
+        if alternative_value is None:
+            alternative_value = Parameter(name, Parameter.Type.NOT_SET)
+
+        return self._parameters.get(name, alternative_value)
 
     def set_parameters(self, parameter_list: List[Parameter]) -> List[SetParametersResult]:
         """
-        Set parameters for the node.
+        Set parameters for the node, and return the result for the set action.
+
+        If any parameter in the list was not declared beforehand and undeclared parameters are not
+        allowed for the node, this method will raise a ParameterNotDeclaredException exception.
+
+        Parameters are set in the order they are declared in the list.
+        If setting a parameter fails due to not being declared, then the
+        parameters which have already been set will stay set, and no attempt will
+        be made to set the parameters which come after.
+
+        If undeclared parameters are allowed, then all the parameters will be implicitly
+        declared before being set even if they were not declared beforehand.
 
         If a callback was registered previously with :func:`set_parameters_callback`, it will be
-        called prior to setting the parameters for the node.
+        called prior to setting the parameters for the node once for each parameter.
+        If the callback prevents a parameter from being set, then it will be reflected in the
+        returned result; no exceptions will be raised in this case.
         For each successfully set parameter, a :class:`ParameterEvent` message is
         published.
 
+        If the value type of the parameter is NOT_SET, and the existing parameter type is
+        something else, then the parameter will be implicitly undeclared.
+
         :param parameter_list: The list of parameters to set.
-        :return: A list of SetParametersResult messages.
+        :return: The result for each set action as a list.
+        :raises: ParameterNotDeclaredException if undeclared paramers are not allowed, and at least
+            one parameter in the list hadn't been declared beforehand.
         """
+        return self._set_parameters(parameter_list, self.set_parameters_atomically)
+
+    def _set_parameters(
+            self,
+            parameter_list: List[Parameter],
+            setter_func=None,
+            raise_on_failure=False
+    ) -> List[SetParametersResult]:
+        """
+        Set parameters for the node, and return the result for the set action.
+
+        Method for internal usage; applies a setter method for each parameters in the list.
+        By default it doesn't check if the parameters were declared, and both declares and sets
+        the given list.
+
+        If a callback was registered previously with :func:`set_parameters_callback`, it will be
+        called prior to setting the parameters for the node once for each parameter.
```

Comment:
@jubeira consider
```suggestion
        called prior to setting the parameters for the node, once for each parameter.
```

Reply:
Addressed in https://github.com/ros2/rclpy/pull/325/commits/4d2429d48296b76742448dccf74d8f4450c43ca9.